apps/src/lib/kits/maker/CircuitPlaygroundBoard.js

```diff
@@ -385,9 +385,49 @@ export default class CircuitPlaygroundBoard extends EventEmitter {
       ? SerialPort
       : ChromeSerialPort.SerialPort;
 
-    return new SerialPortType(portName, {
+    const port = new SerialPortType(portName, {
       baudRate: SERIAL_BAUD
     });
+
+    if (isNodeSerialAvailable()) {
+      const queue = [];
+      let sendPending = false;
+      const oldWrite = port.write;
+
+      // eslint-disable-next-line no-inner-declarations
+      function trySend(buffer) {
+        if (buffer) {
+          queue.push(buffer);
+        }
+
+        if (sendPending || queue.length === 0) {
+          // Exhausted pending send buffer.
+          return;
+        }
+
+        if (queue.length > 512) {
+          throw new Error(
+            'Send queue is full! More than 512 pending messages.'
+          );
+        }
+
+        const toSend = queue.shift();
+        sendPending = true;
+        oldWrite.call(port, toSend, 'binary', function() {
+          sendPending = false;
+
+          if (queue.length !== 0) {
+            trySend();
+          }
+        });
+      }
+
+      port.write = function() {
+        trySend.apply(port, arguments);
```

Comment:
Why `apply` instead of `port.trySend()`?

Reply:
Whoops, I don't think this is needed.

Reply:
Ah — needed to pass along the arguments. Maybe clearer as:

```js
        port.write = (...args) => trySend(...args);
```

Reply:
Don't we know what the arguments are, though? Why not: `port.write = (data, encoding, callback) => trySend(data, encoding, callback)`? Is it because that adds implementation details of `port.write` to CircuitPlaygroundBoard?

Reply:
I think either is prob fine (no strong preference).